source/extensions/common/wasm/wasm.cc

```diff
@@ -495,22 +505,50 @@ static void createWasmInternal(const VmConfig& vm_config, PluginSharedPtr plugin
                  .value_or(code.empty() ? EMPTY_STRING : INLINE_STRING);
   }
 
-  auto callback = [wasm, plugin, cb, source, allow_precompiled = vm_config.allow_precompiled(),
+  auto callback = [vm_config, scope, &cluster_manager, &dispatcher, plugin, cb, source,
                    context_ptr = root_context_for_testing ? root_context_for_testing.release()
                                                           : nullptr](const std::string& code) {
     std::unique_ptr<Context> context(context_ptr);
     if (code.empty()) {
       throw WasmException(fmt::format("Failed to load WASM code from {}", source));
     }
-    if (!wasm->wasm()->initialize(code, allow_precompiled)) {
+    // Construct a unique identifier for the VM based on a hash of the code, vm configuration data
+    // and vm_id.
+    std::string vm_key = Sha256(vm_config.vm_id());
+    vm_key = Xor(vm_key, Sha256(vm_config.configuration()));
+    vm_key = Xor(vm_key, Sha256(code));
+    vm_key = Base64::encode(&*vm_key.begin(), vm_key.size());
+
+    std::shared_ptr<WasmHandle> wasm;
+    {
+      absl::MutexLock l(&base_wasms_mutex_);
+      if (!base_wasms_) {
+        base_wasms_ = new std::remove_reference<decltype(*base_wasms_)>::type;
+      }
+      auto it = base_wasms_->find(vm_key);
+      if (it != base_wasms_->end()) {
+        wasm = it->second.lock();
+        if (!wasm) {
+          base_wasms_->erase(it);
+        }
+      }
+      if (!wasm) {
+        wasm = std::make_shared<WasmHandle>(std::make_shared<Wasm>(
+            vm_config.runtime(), vm_config.vm_id(), vm_config.configuration(), vm_key, scope,
+            cluster_manager, dispatcher));
```

Comment:
Can this ever return `nullptr`? Do we need to handle it or are you following the "fail/crash-fast" principles here?

Reply:
make_shared is never going to return nullptr, and if we OOM, well, we fail fast.